Refactors Mech Recharger

GeneralChaos81 opened the pull request:
Ports TG's mechbay recharger stuff.

Adds the ability to create mech bay recharger turf from 2 plasteel sheets, though maybe I can get rid of that completely.

:cl: General Chaos
add: Ports TG's mech bay recharger code.
/:cl:

Fox-McCloud commented:
Not sure how well this will work. Mech bay rechargers and their associate turfs self reference each other, so I don't think you can just plop one down and it'll be ok.

Honestly, our previous rechargers were better (which checked for mech adjacency)---teh current ones have...issues.

GeneralChaos81 commented:
I've tested it. Putting down the turf works perfectly. The create and destroy procs are already there in code\game\mecha\mech_bay.dm
Tested all the final components: Machine, turf, console. Each of the updates the console and allows the set to function.
Disassembling the machine causes the console to report the failure, so that works too.
Just not crowbarring it, because it doesn't call /turf/simulated/floor/mech_bay_recharge_floor/Destroy() --- I think.

GeneralChaos81 commented:
Could this all be put under '/turf/simulated/floor/mech_bay_recharge_floor/attackby()' work? Or would it be overridden by /turf/simulated/floor/attackby() ?

Fox-McCloud commented:
I'd honestly prefer a refactor of the mech recharging port rather than this.

There's no reason to have a special snowflaked turf when it can all be handled on the mech recharger itself.

GeneralChaos81 commented:
Alright! I did it! Kinda.

- The disassemble is now in mech_bay.dm
It can't be a crowbar attack, I think. Floor.dm's one seems to intercept the call. 
- Does not break screwdrivering wooden floors.
- Proc now calls Destroy(), which disables the attached machines (Woo, that was easy!)
- Still doesn't give you back your sheets. How can I get it to drop some plasteel sheets?

GeneralChaos81 commented:
Thanks to @monster860 and @Crazylemon64 for hints.

- Uses qdel() instead of Destroy() in attackby
- Creates replacement plasteel where you disassembled the plate (instead of somewhere random in the Area)

BeforeChange() sounds to be a better way to call destroying the mechbay machinery than Destroy(), but that won't compile. I get "BeforeChange: undefined proc". You can see how I've attempted to call it commented out in the files changed.
And... if the plate is replaced with a generic tile it is tilepainted to look like it's correct, when it's not.

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -22,6 +25,22 @@
+	if(intact && istype(C, /obj/item/weapon/screwdriver)) //Can't be a crowbar, because otherwise the attackby() for floor.dm intercepts the call, I think
+		to_chat(user, "<span class='danger'>You disassemble the mech bay recharge station.</span>")
+		new /obj/item/stack/sheet/plasteel(src, 2)
+		qdel() //calls the proc above that stops the machine from working - Not required with BeforeChange, but that isn't working.
```
Wait what

<a href=https://github.com/ParadiseSS13/Paradise/blob/master/code/modules/garbage_collection/garbage_collector.dm#L91-L93>I'm not sure qdel does anything without an argument, bucko</a>

GeneralChaos81 replied:
Then it is really weird that it works. And not without. Don't even need this with BeforeChange.

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -12,6 +12,9 @@
 
+
+///turf/simulated/floor/mech_bay_recharge_floor/BeforeChange() //I'm told this is a better thing to call this request by. Undefinied proc error though.
+
```
`BeforeChange` is defined <a href=https://github.com/ParadiseSS13/Paradise/blob/master/code/game/turfs/turf.dm#L168>here</a>. Make sure you've updated your branch to master by doing `git pull upstream master` first, as this change is rather recent.

GeneralChaos81 replied:
Done and working.

Crazylemon64 replied:
You DID do a parent call in that override, right?

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -22,6 +25,22 @@
+		return 1
+	if(intact && istype(C, /obj/item/weapon/screwdriver)) //Can't be a crowbar, because otherwise the attackby() for floor.dm intercepts the call, I think
+		to_chat(user, "<span class='danger'>You disassemble the mech bay recharge station.</span>")
+		new /obj/item/stack/sheet/plasteel(src, 2)
```
The plasteel belongs in an override of `make_plating` - that way, you don't even need to override `attackby`

GeneralChaos81 replied:
The mechbay turf doesn't have builtin_tile (or it does? I'm not sure), so can't be crowbar'd normally. I either need to set that flag, or have the attackby override.
I don't know what best practice is. Should I have the make_tile override anyway?

Crazylemon64 replied:
Did you declare `make_plating` as `floor/mech_bay_recharge_floor/*proc*/make_plating`? The proc term there will cause that - remove it, and it should override fine.

Fox-McCloud commented:
This still isn't acceptable--as I said, utilize TG's recharger, which is far more simple than this and garbage collects a lot easier.

GeneralChaos81 commented:
Riiight. I didn't read between the lines to go get TG's. :P

Crazylemon64 commented:
@Fox-McCloud https://github.com/tgstation/tgstation/blob/master/code/game/mecha/mech_bay.dm I fail to see how -tg- has a simpler recharger.

GeneralChaos81 commented:
Because it has almost half as much code.
https://github.com/ParadiseSS13/Paradise/blob/master/code/game/mecha/mech_bay.dm

Fox-McCloud commented:
@Crazylemon64 

It relies on an oudated global iterator datum: https://github.com/ParadiseSS13/Paradise/blob/a3fdeaa57cdf5b84dc3be2057c0eb0373a41352a/code/datums/helper_datums/global_iterator.dm

It has a lot of cross-references when merely having the charger pick the turf is more than enough: https://github.com/ParadiseSS13/Paradise/blob/master/code/game/mecha/mech_bay.dm#L5

A simple reference from computer to charger and the charger handling which turf can recharge is far more preferable--it also means that you can build mech rechargers anywhere without having to also put in special turfs (that have no reason to exist), not to mention snowflaked code just to make them buildable/deconstructable

GeneralChaos81 commented:
You tricked me! It's awful porting things!!! D:

Crazylemon64 commented:
@GeneralChaos81 The turfs are used in the maps - if you remove the turfs, then the map can't actually compile

GeneralChaos81 commented:
What I mean, was I loaded it up and constructed a console and recharger machine... and they didn't work, until I put down a recharger turf. I figured I made a mistake somewhere, killed the server then went and looked in the code... to find the no references.

My in-game error checking is limited by the fact I can't see what the console is outputting because the UI isn't updated yet. Considering the console doesn't appear to even sample to detect the floor, I don't know that would even help.

Fox-McCloud commented on code/modules/admin/admin.dm:
```diff
@@ -897,8 +897,8 @@ var/gamma_ship_location = 1 // 0 = station , 1 = space
-	for(var/turf/simulated/floor/mech_bay_recharge_floor/F in toArea)
-		F.init_devices()
+//	for(var/turf/simulated/floor/mech_bay_recharge_floor/F in toArea)
+//		F.init_devices()
```
don't add in commented out code.

Fox-McCloud commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -112,7 +112,8 @@ var/global/list/datum/stack_recipe/plasteel_recipes = list(
 	new /datum/stack_recipe("Metal crate", /obj/structure/closet/crate, 10, time = 50, one_per_turf = 1),
-	new /datum/stack_recipe("Mass Driver frame", /obj/machinery/mass_driver_frame, 3, time = 50, one_per_turf = 1)
+	new /datum/stack_recipe("Mass Driver frame", /obj/machinery/mass_driver_frame, 3, time = 50, one_per_turf = 1),
+    new /datum/stack_recipe("Mech Recharge Bay", /turf/simulated/floor/mech_bay_recharge_floor, 2, time = 50, one_per_turf = 1),
```
No longer needed; please remove.

Fox-McCloud commented on code/game/mecha/mech_bay.dm:
```diff
@@ -2,83 +2,26 @@
-	var/datum/global_iterator/mech_bay_recharger/pr_recharger
+	var/max_charge = 50
+	var/on = 0
+	var/repairability = 0 //What's this? I don't think this is a thing on paradise
```
remove this; it's unused on TG and here both.

Fox-McCloud commented:
Make sure this is in attackby

```
	if(default_change_direction_wrench(user, I))
		recharging_turf = get_step(loc, dir)
		return
```

Fox-McCloud commented on code/game/mecha/mech_bay.dm:
```diff
@@ -132,70 +64,49 @@
+		if(recharging_mecha.loc != recharging_turf)
+			recharging_mecha = null
+			recharge_console.update_icon()
+	return //This was missing... Is it not needed? Appeared in the original, but not in the port. Added it back in, to be safe.
```
trailing return. Please remove

Fox-McCloud commented on code/game/mecha/mech_bay.dm:
```diff
@@ -293,3 +157,9 @@
+
+/obj/machinery/computer/mech_bay_power_console/initialize()
+	reconnect()
+	update_icon() //this the right place for it? Otherwise the computer sits there with a black screen.
```
Yes, this is fine. No need for the comment.

Fox-McCloud commented:
Make sure this is in the `attackby`

```
	if(default_change_direction_wrench(user, I))
		recharging_turf = get_step(loc, dir)
		return
```

GeneralChaos81 commented:
Will do. I won't have a chance to change and test stuff until Monday.

GeneralChaos81 commented:
Alright, so the recharger works without the turf. The turf does let you have Charger <> Turf <> Console, where if I removed the turf type would stop that from working. You still can't build or remove the special turf, but at least you can replace mech chargers now.

Added the rotation line in.
Got rid of comments requested.
Got rid of trailing Return.
Still figuring out exactly what data to reference in the nano UI template.

GeneralChaos81 commented:
NanoUI updated and tested successfully.
Got rid of the turf creation line.

I think it's all done.

GeneralChaos81 commented:
@Fox-McCloud I think this is almost ready to go. 
I don't know what I can do about the port destruction not updating the computer though.

Crazylemon64 commented:
Have the port save a reference to the computer, and have it notify the computer on `destroy`?

GeneralChaos81 commented:
I think it has one. `var/obj/machinery/mech_bay_recharge_port/recharge_port`
How would I notify `on destroy`?

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -110,92 +51,61 @@
-/obj/machinery/mech_bay_recharge_port/Destroy()
-	qdel(pr_recharger)
-	pr_recharger = null
-	recharge_console.recharge_port = null
```
```
if(recharge_console)
  recharge_console.recharge_port = null
```

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -206,82 +116,36 @@
 	var/obj/machinery/mech_bay_recharge_port/recharge_port
 
-/obj/machinery/computer/mech_bay_power_console/Destroy()
-	recharge_port.recharge_console = null
```
```
if(recharge_port)
  recharge_port.recharge_console = null
```

Crazylemon64 commented:
How to do so is detailed in the comments I left.

GeneralChaos81 commented:
I did it wrong, and it doesn't work still, but I noticed if you rebuild the machines in their original places they no longer work.
Why do they ever work. What wizardry!?

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -110,92 +51,61 @@
-		return 1
-	else
-		return 0
+/obj/machinery/computer/mech_bay_power_console/proc/reconnect()
```
This should go below where the `mech_bay_power_console` is defined, for order's sake

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -206,82 +116,36 @@
-	if(!recharge_floor || !recharge_port)
-		var/turf/simulated/floor/mech_bay_recharge_floor/F = locate() in range(1,src)
-		if(F)
-			F.init_devices()
```
This should probably call `reconnect` if `recharge_port` is null

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -110,92 +51,61 @@
+	if(recharge_port)
+		return
+	recharge_port = locate(/obj/machinery/mech_bay_recharge_port) in range(1)
+	if(!recharge_port )
```
Fix the spacing here please

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -110,92 +51,61 @@
+			A = get_step(A, D)
+			recharge_port = locate(/obj/machinery/mech_bay_recharge_port) in A
+			if(recharge_port)
+				break
```
You should check that the recharge port doesn't already have a console before you break out of the loop

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -293,3 +157,8 @@
 		ui.set_auto_update(1)
+	return data
+
+/obj/machinery/computer/mech_bay_power_console/initialize()
```
`initialize` should do a parent call, or things will break

Fox-McCloud replied:
Need to follow @Crazylemon64 here, or yeah, it'll be broken.

GeneralChaos81 commented:
- Updated definition order
- Removed extra space
- Killing the port while charging is now reflected on console sprite
- Destroy() now disassociate the other parts.

Still working on stopping two consoles connecting to one port.

Fox-McCloud commented on code/game/mecha/mech_bay.dm:
```diff
@@ -1,84 +1,26 @@
-/turf/simulated/floor/mech_bay_recharge_floor
+/turf/simulated/floor/mech_bay_recharge_floor //I'll remove this in a future PR, as it requires map changes.
```
Don't add in comments like this, and there's nothing wrong with this turf, specifically, anyway.

Crazylemon64 replied:
Besides that it's an artifact of an obsoleted system?

Fox-McCloud commented:
@GeneralChaos81 still need to call `..()` in `initialize`

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -293,3 +173,9 @@
+/obj/machinery/computer/mech_bay_power_console/initialize()
+	reconnect()
+	update_icon()
+	return ..()
```
Nope, it's here

Fox-McCloud commented on code/game/mecha/mech_bay.dm:
```diff
@@ -105,97 +46,55 @@
-		return 0
+	if(default_deconstruction_crowbar(I))
+		recharge_console.recharge_port = null
+		recharge_console.update_icon()
```
The two above are unecessary. When you deconstruct the machine, it completely deletes/destroys it--these procs are called in `Destroy`

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -105,97 +46,53 @@
+	if(!recharging_mecha)
+		recharging_mecha = locate(/obj/mecha) in recharging_turf
+		if(recharging_mecha)
+			recharge_console.update_icon()
```
This should probably display the message above when stepping on a charger turf with a mech

Crazylemon64 replied:
You can put that proc right here

Fox-McCloud replied:
Wouldn't spamming the user with a message every two seconds be a bit annoying though? Not to mention that if power is out, it'll be obvious by everything else being dark.

Crazylemon64 replied:
Nevermind, just use the message displayed in that proc above

Crazylemon64 replied:
It's not gonna spam it, this only gets called if there's no `charging_mech` set - meaning, there was no mech before, but now there is one

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -105,97 +46,53 @@
-			return 1
-		else
-			return 0
-
```
This could stay in, just so that the occupant of the mech can see that the mech is charging properly

GeneralChaos81 replied:
The console starts flashing and stops flashing when the mech starts and stops charging.

Crazylemon64 commented on code/game/mecha/mech_bay.dm:
```diff
@@ -105,97 +46,53 @@
-	else
-		port.stop_charge()
-	return
+			recharge_console.update_icon()
```
Should notify the pilot of the mech when the charging is done

Fox-McCloud commented:
Yeah, I'm not concerned about the messages since there's visual feedback on the monitor

Crazylemon64 commented:
I guess that works, then

Fox-McCloud commented:
Going to go ahead and merge this; tested it and it works fine with NanoUI, GC's, doesn't runtime, and functions properly.
